File: proposed-projects/round-trip-experiment/reverse/20260909T011452Z_9f286468/20260910T045943Z_8c3d5bc1/ported_implementation/src/rules-builder/rules-builder.ts
import { GrammarParseError } from '../utils/errors/grammar-parse-error';
import { isWordChar } from './is-word-char';
import { parseChar } from './parse-char';
import { parseName } from './parse-name';
import { parseSpace } from './parse-space';
import { SymbolIds } from './symbol-ids';
import { InternalRuleDef, InternalRuleType } from './types';

const WHITESPACE = /\s/;

const now = (): number => performance.now();

export class RulesBuilder {
  public pos = 0;
  public symbolIds = new SymbolIds();
  public rules: (InternalRuleDef[] | undefined)[] = [];
  public src: string;
  public start: number;
  private timeLimit: number;

  constructor(src: string, limit = 1000) {
    this.src = src;
    this.start = now();
    this.timeLimit = limit;
    this.parse(src);
  }

  private parse(src: string): void {
    // move cursor forward until we reach non-whitespace content
    this.pos = parseSpace(src, 0, true);

    while (this.pos < src.length) {
      this.parseRule(src);
    }

    // Validate the state to ensure that all rules are defined
    for (const rule of this.rules) {
      if (rule === undefined) {
        continue;
      }
      for (const elem of rule) {
        if (elem.type === InternalRuleType.RULE_REF) {
          // Ensure that the rule at that location exists
          const value = elem.value as number;
          const referenced = this.rules[value];
          const ruleExists =
            value < this.rules.length &&
            referenced !== undefined &&
            referenced.length > 0;
          if (!ruleExists) {
            const missingRuleName = this.symbolIds.reverseGet(value);
            let missingRulePos = this.symbolIds.getPos(missingRuleName);

            // Skip over the ::= and any whitespace
            while (
              missingRulePos < src.length &&
              (src[missingRulePos] === ':' ||
                src[missingRulePos] === '=' ||
                WHITESPACE.test(src[missingRulePos]))
            ) {
              missingRulePos += 1;
            }

            throw new GrammarParseError(
              src,
              missingRulePos,
              `Undefined rule identifier "${missingRuleName}"`
            );
          }
        }
      }
    }
  }

  private parseRule(src: string): void {
    const name = parseName(src, this.pos);
    this.pos = parseSpace(src, this.pos + name.length, false);
    const ruleId = this.getSymbolId(name, name.length);

    // Skip over whitespace characters and find the ::= sequence
    this.pos = parseSpace(src, this.pos, true);
    if (
      !(
        src[this.pos] === ':' &&
        src[this.pos + 1] === ':' &&
        src[this.pos + 2] === '='
      )
    ) {
      throw new GrammarParseError(src, this.pos, `Expecting ::= at ${this.pos}`);
    }
    this.pos = parseSpace(src, this.pos + 3, true);

    this.parseAlternates(name, ruleId);

    if (src[this.pos] === '\r') {
      this.pos += src[this.pos + 1] === '\n' ? 2 : 1;
    } else if (src[this.pos] === '\n') {
      this.pos += 1;
    } else if (src[this.pos]) {
      throw new GrammarParseError(
        src,
        this.pos,
        `Expecting newline or end at ${this.pos}`
      );
    }
    this.pos = parseSpace(src, this.pos, true);
  }

  private getSymbolId(src: string, length: number): number {
    const nextId = this.symbolIds.size;
    const key = src.slice(0, length);
    if (!this.symbolIds.has(key)) {
      this.symbolIds.set(key, nextId, this.pos);
    }
    return this.symbolIds.get(key);
  }

  private generateSymbolId(baseName: string): number {
    const nextId = this.symbolIds.size;
    this.symbolIds.set(`${baseName}_${nextId}`, nextId, this.pos);
    return nextId;
  }

  private addRule(ruleId: number, rule: InternalRuleDef[]): void {
    while (this.rules.length <= ruleId) {
      this.rules.push(undefined);
    }
    this.rules[ruleId] = rule;
  }

  private checkDuration(): void {
    if (now() - this.start > this.timeLimit) {
      throw new GrammarParseError(
        this.src,
        this.pos,
        `duration of ${this.timeLimit} exceeded:`
      );
    }
  }

  private parseSequence(
    ruleName: string,
    outElements: InternalRuleDef[],
    depth = 0
  ): void {
    const isNested = depth !== 0;
    const src = this.src;
    let lastSymStart = outElements.length;
    while (src[this.pos]) {
      if (src[this.pos] === '"') {
        this.pos += 1;
        lastSymStart = outElements.length;
        while (src[this.pos] !== '"') {
          this.checkDuration();
          const [value, incPos] = parseChar(src, this.pos);
          outElements.push(new InternalRuleDef(InternalRuleType.CHAR, [value]));
          // Adjusting pos by the length of parsed characters
          this.pos += incPos;
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (src[this.pos] === '[') {
        this.pos += 1;
        let startType = InternalRuleType.CHAR;
        if (src[this.pos] === '^') {
          this.pos += 1;
          startType = InternalRuleType.CHAR_NOT;
        }
        lastSymStart = outElements.length;
        while (src[this.pos] !== ']') {
          this.checkDuration();
          const type =
            lastSymStart < outElements.length
              ? InternalRuleType.CHAR_ALT
              : startType;
          const [startcharValue, incPos] = parseChar(src, this.pos);
          this.pos += incPos;
          if (
            type === InternalRuleType.CHAR ||
            type === InternalRuleType.CHAR_NOT
          ) {
            outElements.push(new InternalRuleDef(type, [startcharValue]));
          } else {
            outElements.push(new InternalRuleDef(type, startcharValue));
          }

          if (src[this.pos] === '-' && src[this.pos + 1] !== ']') {
            this.pos += 1;
            const [endcharValue, endIncPos] = parseChar(src, this.pos);
            outElements.push(
              new InternalRuleDef(
                InternalRuleType.CHAR_RNG_UPPER,
                endcharValue
              )
            );
            this.pos += endIncPos;
          }
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (isWordChar(src[this.pos])) {
        const name = parseName(src, this.pos);
        const refRuleId = this.getSymbolId(name, name.length);
        this.pos += name.length;
        this.pos = parseSpace(src, this.pos, isNested);

        lastSymStart = outElements.length;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, refRuleId)
        );
      } else if (src[this.pos] === '(') {
        this.pos = parseSpace(src, this.pos + 1, true);
        const subRuleId = this.generateSymbolId(ruleName);
        this.parseAlternates(ruleName, subRuleId, depth + 1);
        lastSymStart = outElements.length;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId)
        );
        if (src[this.pos] !== ')') {
          throw new GrammarParseError(
            src,
            this.pos,
            `Expecting ')' at ${this.pos}`
          );
        }
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else if (
        src[this.pos] === '*' ||
        src[this.pos] === '+' ||
        src[this.pos] === '?'
      ) {
        if (lastSymStart === outElements.length) {
          throw new GrammarParseError(
            src,
            this.pos,
            `Expecting preceding item to */+/? at ${this.pos}`
          );
        }
        const subRuleId = this.generateSymbolId(ruleName);
        const subRule: InternalRuleDef[] = outElements.slice(lastSymStart);
        if (src[this.pos] === '*' || src[this.pos] === '+') {
          subRule.push(new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId));
        }
        subRule.push(new InternalRuleDef(InternalRuleType.ALT));
        if (src[this.pos] === '+') {
          subRule.push(...outElements.slice(lastSymStart));
        }
        subRule.push(new InternalRuleDef(InternalRuleType.END));
        this.addRule(subRuleId, subRule);
        outElements.length = lastSymStart;
        outElements.push(
          new InternalRuleDef(InternalRuleType.RULE_REF, subRuleId)
        );
        this.pos = parseSpace(src, this.pos + 1, isNested);
      } else {
        break;
      }
    }
  }

  private parseAlternates(ruleName: string, ruleId: number, depth = 0): void {
    const src = this.src;
    const rule: InternalRuleDef[] = [];
    this.parseSequence(ruleName, rule, depth);
    while (src[this.pos] === '|') {
      this.checkDuration();
      rule.push(new InternalRuleDef(InternalRuleType.ALT));
      this.pos = parseSpace(src, this.pos + 1, true);
      this.parseSequence(ruleName, rule, depth);
    }
    rule.push(new InternalRuleDef(InternalRuleType.END));
    this.addRule(ruleId, rule);
  }
}

File: packages/gbnf-experiment/data/20260909T032340Z_f01d5496/ported_implementation/gbnf/rules_builder/rules_builder.py
import re
import time
from typing import Dict, List, Optional

from ..utils.errors.grammar_parse_error import GrammarParseError
from .is_word_char import is_word_char
from .parse_char import parse_char
from .parse_name import parse_name
from .parse_space import parse_space
from .symbol_ids import SymbolIds
from .types import InternalRuleDef, InternalRuleType

_WHITESPACE = re.compile(r"\s")


class RulesBuilder:
    def __init__(self, src: str, limit: int = 1000):
        self.pos = 0
        self.symbol_ids = SymbolIds()
        # rules are assigned by id, and ids are not necessarily assigned in order, so
        # they're collected in a map and flattened out into a (possibly sparse) list.
        self._rules_by_id: Dict[int, List[InternalRuleDef]] = {}
        self.src = src
        self.start = time.perf_counter() * 1000
        self._time_limit = limit
        self.parse(src)

    @property
    def rules(self) -> List[Optional[List[InternalRuleDef]]]:
        if not self._rules_by_id:
            return []
        length = max(self._rules_by_id) + 1
        return [self._rules_by_id.get(idx) for idx in range(length)]

    def _at(self, pos: int):
        return self.src[pos] if 0 <= pos < len(self.src) else None

    def parse(self, src: str) -> None:
        # move cursor forward until we reach non-whitespace content
        self.pos = parse_space(src, 0, True)

        while self.pos < len(src):
            self.parse_rule(src)

        # Validate the state to ensure that all rules are defined
        for rule in self.rules:
            if rule is None:
                continue
            for elem in rule:
                if elem.type == InternalRuleType.RULE_REF:
                    # Ensure that the rule at that location exists
                    referenced = self._rules_by_id.get(elem.value)
                    rule_exists = referenced is not None and len(referenced) > 0
                    if not rule_exists:
                        missing_rule_name = self.symbol_ids.reverse_get(elem.value)
                        missing_rule_pos = self.symbol_ids.get_pos(missing_rule_name)

                        # Skip over the ::= and any whitespace
                        while missing_rule_pos < len(src) and (
                            src[missing_rule_pos] in (":", "=")
                            or _WHITESPACE.search(src[missing_rule_pos])
                        ):
                            missing_rule_pos += 1

                        raise GrammarParseError(
                            src,
                            missing_rule_pos,
                            f'Undefined rule identifier "{missing_rule_name}"',
                        )

    def parse_rule(self, src: str) -> None:
        name = parse_name(src, self.pos)
        self.pos = parse_space(src, self.pos + len(name), False)
        rule_id = self.get_symbol_id(name, len(name))

        # Skip over whitespace characters and find the ::= sequence
        self.pos = parse_space(src, self.pos, True)
        if not (
            self._at(self.pos) == ":"
            and self._at(self.pos + 1) == ":"
            and self._at(self.pos + 2) == "="
        ):
            raise GrammarParseError(src, self.pos, f"Expecting ::= at {self.pos}")
        self.pos = parse_space(src, self.pos + 3, True)

        self.parse_alternates(name, rule_id)

        if self._at(self.pos) == "\r":
            self.pos += 2 if self._at(self.pos + 1) == "\n" else 1
        elif self._at(self.pos) == "\n":
            self.pos += 1
        elif self._at(self.pos):
            raise GrammarParseError(
                src, self.pos, f"Expecting newline or end at {self.pos}"
            )
        self.pos = parse_space(src, self.pos, True)

    def get_symbol_id(self, src: str, length: int) -> int:
        next_id = self.symbol_ids.size
        key = src[0:length]
        if not self.symbol_ids.has(key):
            self.symbol_ids.set(key, next_id, self.pos)
        return self.symbol_ids.get(key)

    def generate_symbol_id(self, base_name: str) -> int:
        next_id = self.symbol_ids.size
        self.symbol_ids.set(f"{base_name}_{next_id}", next_id, self.pos)
        return next_id

    def add_rule(self, rule_id: int, rule: List[InternalRuleDef]) -> None:
        self._rules_by_id[rule_id] = rule

    def check_duration(self) -> None:
        if time.perf_counter() * 1000 - self.start > self._time_limit:
            raise GrammarParseError(
                self.src, self.pos, f"duration of {self._time_limit} exceeded:"
            )

    def parse_sequence(
        self,
        rule_name: str,
        out_elements: List[InternalRuleDef],
        depth: int = 0,
    ) -> None:
        is_nested = depth != 0
        src = self.src
        last_sym_start = len(out_elements)
        while self._at(self.pos):
            char = self._at(self.pos)
            if char == '"':
                self.pos += 1
                last_sym_start = len(out_elements)
                while self._at(self.pos) != '"':
                    self.check_duration()
                    value, inc_pos = parse_char(src, self.pos)
                    out_elements.append(
                        InternalRuleDef(InternalRuleType.CHAR, [value])
                    )
                    # Adjusting pos by the length of parsed characters
                    self.pos += inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif char == "[":
                self.pos += 1
                start_type = InternalRuleType.CHAR
                if self._at(self.pos) == "^":
                    self.pos += 1
                    start_type = InternalRuleType.CHAR_NOT
                last_sym_start = len(out_elements)
                while self._at(self.pos) != "]":
                    self.check_duration()
                    type_ = (
                        InternalRuleType.CHAR_ALT
                        if last_sym_start < len(out_elements)
                        else start_type
                    )
                    startchar_value, inc_pos = parse_char(src, self.pos)
                    self.pos += inc_pos
                    if type_ in (InternalRuleType.CHAR, InternalRuleType.CHAR_NOT):
                        out_elements.append(
                            InternalRuleDef(type_, [startchar_value])
                        )
                    else:
                        out_elements.append(
                            InternalRuleDef(type_, startchar_value)
                        )

                    if self._at(self.pos) == "-" and self._at(self.pos + 1) != "]":
                        self.pos += 1
                        endchar_value, inc_pos = parse_char(src, self.pos)
                        out_elements.append(
                            InternalRuleDef(
                                InternalRuleType.CHAR_RNG_UPPER, endchar_value
                            )
                        )
                        self.pos += inc_pos
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif is_word_char(char):
                name = parse_name(src, self.pos)
                ref_rule_id = self.get_symbol_id(name, len(name))
                self.pos += len(name)
                self.pos = parse_space(src, self.pos, is_nested)

                last_sym_start = len(out_elements)
                out_elements.append(
                    InternalRuleDef(InternalRuleType.RULE_REF, ref_rule_id)
                )
            elif char == "(":
                self.pos = parse_space(src, self.pos + 1, True)
                sub_rule_id = self.generate_symbol_id(rule_name)
                self.parse_alternates(rule_name, sub_rule_id, depth + 1)
                last_sym_start = len(out_elements)
                out_elements.append(
                    InternalRuleDef(InternalRuleType.RULE_REF, sub_rule_id)
                )
                if self._at(self.pos) != ")":
                    raise GrammarParseError(
                        src, self.pos, f"Expecting ')' at {self.pos}"
                    )
                self.pos = parse_space(src, self.pos + 1, is_nested)
            elif char in ("*", "+", "?"):
                if last_sym_start == len(out_elements):
                    raise GrammarParseError(
                        src,
                        self.pos,
                        f"Expecting preceding item to */+/? at {self.pos}",
                    )
                sub_rule_id = self.generate_symbol_id(rule_name)
                sub_rule: List[InternalRuleDef] = out_elements[last_sym_start:]
                if char in ("*", "+"):
                    sub_rule.append(
                        InternalRuleDef(InternalRuleType.RULE_REF, sub_rule_id)
                    )
                sub_rule.append(InternalRuleDef(InternalRuleType.ALT))
                if char == "+":
                    sub_rule.extend(out_elements[last_sym_start:])
                sub_rule.append(InternalRuleDef(InternalRuleType.END))
                self.add_rule(sub_rule_id, sub_rule)
                del out_elements[last_sym_start:]
                out_elements.append(
                    InternalRuleDef(InternalRuleType.RULE_REF, sub_rule_id)
                )
                self.pos = parse_space(src, self.pos + 1, is_nested)
            else:
                break

    def parse_alternates(
        self,
        rule_name: str,
        rule_id: int,
        depth: int = 0,
    ) -> None:
        src = self.src
        rule: List[InternalRuleDef] = []
        self.parse_sequence(rule_name, rule, depth)
        while self._at(self.pos) == "|":
            self.check_duration()
            rule.append(InternalRuleDef(InternalRuleType.ALT))
            self.pos = parse_space(src, self.pos + 1, True)
            self.parse_sequence(rule_name, rule, depth)
        rule.append(InternalRuleDef(InternalRuleType.END))
        self.add_rule(rule_id, rule)
